### ISSUE TITLE
Paket breaks hot reload

### ISSUE THREAD

**Issue by jakkn** [nwn-dotnet/Anvil#451]
Loading plugins with Paket encounters the following during Anvil reload
```log
E [2021/12/31 12:39:55.947] [BoP.HotReload.HotReloadService] Plugin assembly change detected! Triggering reload of Anvil.
I [2021/12/31 12:39:55.973] [Anvil.AnvilCore] Reloading Anvil
I [2021/12/31 12:39:55.974] [Anvil.Services.ServiceManager] Unloading services...
I [2021/12/31 12:39:56.003] [Anvil.Plugins.PluginManager] Unloading plugins...
I [2021/12/31 12:39:56.003] [Anvil.Plugins.PluginManager] Unloading DotNET plugin "NWN.ChatTools" - "/dotNET/bin/Paket/packages/NWN.ChatTools/lib/net5.0/NWN.ChatTools.dll"
```
<details open>
<summary>lldb dumpheap</summary>

```
(lldb) dumpheap -type LoaderAllocator
         Address               MT     Size
00007fd0e4263b50 00007fd11e75de28       48     
00007fd0e4263b80 00007fd11f519a00       24     
00007fd0e43544f0 00007fd11e75de28       48     
00007fd0e4354520 00007fd11f519a00       24     
00007fd0e4354df0 00007fd11e75de28       48     
00007fd0e4354e20 00007fd11f519a00       24     
00007fd0e438a030 00007fd11e75de28       48     
00007fd0e438a060 00007fd11f519a00       24     

Statistics:
              MT    Count    TotalSize Class Name
00007fd11f519a00        4           96 System.Reflection.LoaderAllocatorScout
00007fd11e75de28        4          192 System.Reflection.LoaderAllocator
Total 8 objects

(lldb) gcroot -all 0x00007fd0e4263b50
Thread 1:
    00007FFE9C16A240 00007FD11FAB8710 Anvil.Plugins.PluginManager.Unload() [/home/runner/work/Anvil/Anvil/NWN.Anvil/src/main/Plugins/PluginManager.cs @ 90]
        rbx: 
            ->  00007FD0E400C438 Anvil.Plugins.PluginManager
            ->  00007FD0E400C4B0 System.Collections.Generic.List`1[[Anvil.Plugins.Plugin, NWN.Anvil]]
            ->  00007FD0E4303348 Anvil.Plugins.Plugin[]
            ->  00007FD0E4306480 Anvil.Plugins.Plugin
            ->  00007FD0E4263EA8 System.Reflection.RuntimeAssembly
            ->  00007FD0E4263B50 System.Reflection.LoaderAllocator

    00007FFE9C16A240 00007FD11FAB8710 Anvil.Plugins.PluginManager.Unload() [/home/runner/work/Anvil/Anvil/NWN.Anvil/src/main/Plugins/PluginManager.cs @ 90]
        rbp-28: 00007ffe9c16a248
            ->  00007FD0E400C4B0 System.Collections.Generic.List`1[[Anvil.Plugins.Plugin, NWN.Anvil]]
            ->  00007FD0E4303348 Anvil.Plugins.Plugin[]
            ->  00007FD0E4306480 Anvil.Plugins.Plugin
            ->  00007FD0E4263EA8 System.Reflection.RuntimeAssembly
            ->  00007FD0E4263B50 System.Reflection.LoaderAllocator

HandleTable:
    00007FD1194C1288 (strong handle)
    -> 00007FD0FC008C28 System.Object[]
    -> 00007FD0E4081280 Microsoft.FSharp.Control.FSharpEvent`1[[Paket.Logging+Trace, Paket.Core]]
    -> 00007FD0E40815B0 Microsoft.FSharp.Control.FSharpHandler`1[[Paket.Logging+Trace, Paket.Core]]
    -> 00007FD0E4080B78 Anvil.Plugins.PaketPluginSource
    -> 00007FD0E400C438 Anvil.Plugins.PluginManager
    -> 00007FD0E400C4B0 System.Collections.Generic.List`1[[Anvil.Plugins.Plugin, NWN.Anvil]]
    -> 00007FD0E4303348 Anvil.Plugins.Plugin[]
    -> 00007FD0E4306480 Anvil.Plugins.Plugin
    -> 00007FD0E4263EA8 System.Reflection.RuntimeAssembly
    -> 00007FD0E4263B50 System.Reflection.LoaderAllocator

    00007FD1194C1368 (strong handle)
    -> 00007FD0FC003420 System.Object[]
    -> 00007FD0E4009CB8 Anvil.AnvilCore
    -> 00007FD0E400C438 Anvil.Plugins.PluginManager
    -> 00007FD0E400C4B0 System.Collections.Generic.List`1[[Anvil.Plugins.Plugin, NWN.Anvil]]
    -> 00007FD0E4303348 Anvil.Plugins.Plugin[]
    -> 00007FD0E4306480 Anvil.Plugins.Plugin
    -> 00007FD0E4263EA8 System.Reflection.RuntimeAssembly
    -> 00007FD0E4263B50 System.Reflection.LoaderAllocator

    00007FD1194C4200 (strong handle)
    -> 00007FD0E4354418 Anvil.Plugins.PluginLoadContext
    -> 00007FD0E400C438 Anvil.Plugins.PluginManager
    -> 00007FD0E400C4B0 System.Collections.Generic.List`1[[Anvil.Plugins.Plugin, NWN.Anvil]]
    -> 00007FD0E4303348 Anvil.Plugins.Plugin[]
    -> 00007FD0E4306480 Anvil.Plugins.Plugin
    -> 00007FD0E4263EA8 System.Reflection.RuntimeAssembly
    -> 00007FD0E4263B50 System.Reflection.LoaderAllocator

    00007FD1194C4A48 (strong handle)
    -> 00007FD0E4263B50 System.Reflection.LoaderAllocator

Found 6 roots.
```
</details>

`PaketPluginSource` seems to hold a few resources. This log handler looks like it should be unregistered https://github.com/nwn-dotnet/Anvil/blob/adaf95247ea27c96c53c7d92899bc4e97d480f8b/NWN.Anvil/src/main/Plugins/Paket/PaketPluginSource.cs#L44

I tried making the class `IDisposable` but that didn't work. I guess Paket is never disposed, at least not early enough.

**Comment by Jorteck** [comment]
Can you see if this is resolved from https://github.com/nwn-dotnet/Anvil/pull/453?

Also, are you loading two plugins with the same name from the plugin folder & paket?

**Comment by jakkn** [comment]
> Can you see if this is resolved from #453?

Still fails with the same error.

> Also, are you loading two plugins with the same name from the plugin folder & paket?

Not intentionally - is there anything that makes you think so? `ANVIL_HOME` is set to `/dotNET/bin` in my image, and the binary has been copied from the output dir of a gitlab runner which I believe is why lldb reports paths such as `/home/runner/work/Anvil/Anvil/NWN.Anvil/`

**Comment by Jorteck** [comment]
`Anvil.Plugins.Plugin` should lose the assembly reference immediately after this log message:
```cs
Unloading DotNET plugin "NWN.ChatTools"
```

The dispose call:
https://github.com/nwn-dotnet/Anvil/blob/development/NWN.Anvil/src/main/Plugins/PluginManager.cs#L88-L89

Dispose implementation:
https://github.com/nwn-dotnet/Anvil/blob/development/NWN.Anvil/src/main/Plugins/Plugin.cs#L42

The only edge case I can think of is if this list is populated with two identical entries of the same plugin. Hence my thought process with my last comment :P

Could you try dumping the contents of that list during unload? I can't really think of other places that hold a direct reference.
